### ISSUE TITLE
App router: Content flickering with React.createContext and next/dynamic for SSG/SSR

### ISSUE THREAD

**Issue by KirillEvtushenko** [vercel/next.js#64060]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/interesting-aryabhata-ckyy6f

### To Reproduce

1. Refresh the page
2. Notice that the right block with Client component and RSC inside noticeably flickers

### Current vs. Expected behavior

When using `next/dynamic` with React.createContext, I expect that it will work the same way as without React.createContext.
Please pay attention that initial request with html correctly contains all the needed html. Provider's rerender (recreating of the `value` object prop) then causes that flickering.

https://github.com/vercel/next.js/assets/102147606/165d4731-c2e4-4eb1-94f7-1f6d42898d75

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.4.0: Wed Feb 21 21:44:43 PST 2024; root:xnu-10063.101.15~2/RELEASE_ARM64_T6000
Binaries:
  Node: 20.10.0
  npm: 10.2.3
  Yarn: 1.22.18
  pnpm: 8.6.5
Relevant Packages:
  next: 14.1.4
  eslint-config-next: 13.5.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.4.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Dynamic imports (next/dynamic)

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local), Vercel (Deployed), Other (Deployed)

<sub>[NEXT-3074](https://linear.app/vercel/issue/NEXT-3074/app-router-content-flickering-with-reactcreatecontext-and-nextdynamic)</sub>

**Comment by huozhi** [comment]
Can you try switching `dynamic()` call to `React.lazy()` if you're not using `ssr: false`

**Comment by KirillEvtushenko** [comment]
Hey @huozhi ! Thanks for the answer.
I've tried to switch `dynamic()` calls to `lazy()`:
```ts
'use client';

import { lazy } from 'react';

export const ClientComponent = lazy(() => import('./ClientComponent'));
export const ClientComponent2 = lazy(() => import('./ClientComponent2'));
// Other client components imports
```
However, I get this error (have never seen this kind of error before lol):
`Error: Element type is invalid. Received a promise that resolves to: [object Object]. Lazy element type must resolve to a class or function. Did you wrap a component in React.lazy() more than once?`

I found the only way to fix it: remove `"use client"` directive from the registry. But this breaks [code-splitting and puts everything to the First load JS](https://github.com/vercel/next.js/issues/49454#issuecomment-1830053413)

Am I missing something?

**Comment by joshwcomeau** [comment]
Hi @huozhi!

I'm running into this issue as well, and also thought to try `React.lazy()`. This does indeed solve the flickering issue, but it causes a regression around the CSS being included on SSR; the awesome work you did to close #62940 isn't applied with `React.lazy()`.

But I'm glad to hear that there's a potential fix, with removing the Suspense boundary when no loading state is provided! I believe this would solve the issue 👍.

**Comment by bluebeel** [comment]
I can confirm the next/dynamic doesn't work even if done in a `"use client"` file.
The solution we did for now was using React.lazy where we could. But yeah quite the regression compared to Pages Router ^^'

**Comment by KirillEvtushenko** [comment]
Moving aside the issue with _React.ContextAPI_ breaking _next/dynamic_, there is one more “issue” I want to show.   

We have a number of JAMStack apps built with Next.js, that we are currently moving from Pages router to App router. In CMS we have a content that describes a page -> sections, their order, props etc. Nothing fancy there, quite common approach. To map CMS content to actual components, we use this kind of registry (simplified): 
```js
// registry.ts
const ClientComponentLvl1 = dynamic(() => import('./atoms/ClientComponentLvl1'));
const ClientComponentLvl2 = dynamic(() => import('./molecules/ClientComponentLvl2'));
const ClientComponentLvl3 = dynamic(() => import('./organisms/ClientComponentLvl3'));

const components = {
  ClientComponentLvl1,
  ClientComponentLvl2,
  ClientComponentLvl3
}
export const getComponent = (name: keyof typeof components) => components[name]


// page.tsx
import {getComponent}  from "@/components/registry";

const getPageProps = async () => {
  const props = await fetch(`content from CMS`);
  return props;
}

export default async function Page() {
  const props = await getPageProps();

  const Section = getComponent(props.sectionName);

  return (
    <main>
      <Section />
    </main>
  );
}
```

All pages are built as SSG. If we open some page, HTML loads as it supposed to, we immediately see the content, no CLS. The problem appears if we NAVIGATE to this page using built-in `<Link />` component. We don’t see any HTML for "use client" components then and we suffer from noticeable CLS as dynamic components start actually to load dynamically.

Opened directly (looks good):  

https://github.com/vercel/next.js/assets/102147606/582a74e6-1602-49a1-b979-876083a38ede

Internal navigation (CLS):

https://github.com/vercel/next.js/assets/102147606/8e0703a0-4087-4426-84e2-4da1ee2e9d78


So again: we can directly open the SSG generated page and see the content, but we CAN’T navigate to this page through the app without seeing how it dynamically loads. And this applies only to the ‘use client’ components. RSC is good.
“funny” thing is that the workaround would be to change `<Link />` to `<a />` tag. Full redirect with reloading of html makes it look much better and smoother, ha.
This behaviour seems wrong. Or do we miss some other way of utilizing JAMStack in Next.js.

**Comment by huozhi** [comment]
We landed a fix on canary, can anyone confirm if it fixed the problem for them? ty

**Comment by huozhi** [comment]
In the repro seems there's a setState call in React context Provider when on mount, which will trigger the re-render of Suspense boundary in `next/dynamic` that lead to the flicking you saw. We recommend to remove the setState in the context provider.
The reproduction demostrates the issue well but I wonder what's the real-world case as trigger re-render in context Provider will trigger all the re-render in the Suspense boundaries below.

**Comment by KirillEvtushenko** [comment]
Hi! The real-world use case is loading some initial data from localStorage and setting it as a state on mount. We need it. And I can come up with dozens of other use cases.

**Comment by snehakhobragade90** [comment]
@huozhi Removing of extra suspense boundary fixed the loading state to flash before rendering the next/dynamic - https://github.com/vercel/next.js/issues/64687
Without this fix we are completely blocked to proceed with our app migration. For the initial launch we ended up statically importing the components which are above the pageFold, but we cannot continue doing this as we cannot treeshake this and it abundantly increases the app size.

> If you want to mark this behaviour as "intended", I would like to ask you to help us find any alternatives to "next/dynamic" as we weren't able to achieve the same level of tree shaking with static imports.

💯  💯 💯 
I also wanted to add, in my [sandbox example](https://codesandbox.io/p/devbox/jolly-fog-h86q9z) I have created a minimal code to reproduce the bug. The `setState` call is not in any React contextProvider which you have mentioned in the comment above. But instead in a simple react component which renders a next/dynamic imports. I have more details in added in the issue if it helps to debug it better. Thanks.

**Comment by huozhi** [comment]
Hi @shanekunz the setState in provider was the original reproduction of this issue. In your repro there's a setState call above the `next/dynamic` component which will also trigger the Suspense re-rendering

```
useEffect(() => {
    setsomeState("bar");
  });
```

**Comment by huozhi** [comment]
@KirillEvtushenko I patched an improvement in #65486 that will preload the chunks earlier, and seemed to work for me with the initial reproduction I downloaded before. Could you help verify with latest canary again to see if it helps since I don't have the source code of the demo from the video?

**Comment by nmiddendorff** [comment]
> Hi @snehakhobragade90 the setState in provider was the original reproduction of this issue. In your repro there's a setState call above the next/dynamic component which will also trigger the Suspense re-rendering

@huozhi, that's correct. Calling `useState` directly in the parent OR `setState` in a provider will trigger next/dynamic's [baked-in suspense boundary](https://github.com/vercel/next.js/pull/65309). I don't know if it's realistic to recommend that parents and siblings to a dynamic component should not call `useState` or `setState`. The baked in suspense boundary does have me looking at using `React.lazy()` instead to avoid the unwanted suspense boundary but as @joshwcomeau mentioned this reintroduces the style flashing issue #62940 that was patched for next/dynamic. Is a potential path forward to apply a [similar fix](https://github.com/vercel/next.js/pull/64294) to `React.lazy` so CSS is collected before rendering client side?

**Comment by KirillEvtushenko** [comment]
> @KirillEvtushenko I patched an improvement in #65486 that will preload the chunks earlier, and seemed to work for me with the initial reproduction I downloaded before. Could you help verify with latest canary again to see if it helps since I don't have the source code of the demo from the video?
> 
> Also regarding to the read of localStorage, is that possible to finish in callback of `useState` like `useState(() => { .. })` so you don't need to set it again to trigger the re-render?

@huozhi, tested on `14.3.0-canary.57`. Both issues (on direct opening and on redirect) are still there, unfortunately. I clearly see that waterfall improved and components appear faster, but it's still visually visible. And thats for the **direct** opening process. 

As for **redirect** to the page using Link component, dynamic loading is still consequent:

https://github.com/vercel/next.js/assets/102147606/6325340d-7d18-4f39-8ab9-86f9a3d96ecb

**Comment by huozhi** [comment]
@KirillEvtushenko could you share the codebase of this video to let us dig into it? 🙏

**Comment by KirillEvtushenko** [comment]
Of course! Updated the repo and readme.
https://github.com/KirillEvtushenko/app-router-dynamic-imports-react-context-issue

**Comment by snehakhobragade90** [comment]
@huozhi quick proposal, thoughts if we want to add the extra suspense as a conditional element ? We can probably use the [suspense](https://github.com/vercel/next.js/pull/64716/files#diff-049c589e1af4439782c0011bdcfd90c238a538bed7357eef9d98cd0f74bd737aL46) prop (which I see was deprecated, or we can introduce a new prop to handle this logic)
The default behavior can be to render with the Suspense, but the consumers can opt out of it if required. Thoughts ?

**Comment by brian123zx** [comment]
One more repro case to pile on here. When the initial fix PR was [reverted](https://github.com/vercel/next.js/pull/65309), I dug into the history of `next/dynamic` and saw others reporting similar issues as us on the [PR](https://github.com/vercel/next.js/pull/42589) which introduced its current implementation. The suggested fix was to [wrap state changes in `startTransition`](https://github.com/vercel/next.js/pull/42589#issuecomment-1381006680). One response to that in the thread [suggests](https://github.com/vercel/next.js/pull/42589#issuecomment-1415344561) a problem when the state updates are within a 3rd party hook out of your control. We're now running into exactly this, so I've set up an isolated repro [here](https://codesandbox.io/p/devbox/adoring-liskov-3vxn4y) to illustrate the problem.

**Comment by navin-moorthy** [comment]
Might be related to this issues - https://github.com/vercel/next.js/issues/65796

But the flickering can be seen by doing the below steps,
1. Go to HomePage `/` - https://lh4gk5-3000.csb.app/
2. Go to AboutPage `/about`
3. Hard refresh
4. Go to Homepage '`'
5. To reproduce again, Hard refresh
6. Go to AboutPage `/about`

**Comment by snehakhobragade90** [comment]
@huozhi are there any updates on this ? We are completely blocked to launch our Nextjs project due to the lack of treeshaking support coupled with these unresolved issues. Any updates would be helpful to plan next steps at our end. Thanks

**Comment by huozhi** [comment]
I'm looking into this now with previous approach, but there's still errors could throw on client that breaks the dom insertion. We're still investigating

**Comment by huozhi** [comment]
We've landed #67014 on canary as the fix, please test it out 🙏 The repro works well locally for me. I'll close this for now
